### DISCUSSION TITLE
What would be the best place to attach an internal Client to a server?

### DISCUSSION THREAD

**Discussion by arnoweiss** [eclipse-milo/milo#914] [General]
Hi - I am trying to create an AddressSpace mapper: OPC UA Client on one side, OPC UA Server on the other. The best I've come up with so far is to add the logic to the class spanning the Namespace. I create a separate Thread for each server-side VariableNode to implement the client and mapping logic there - similar to the BogusEventNotifier from the ExampleServer. I feel like it would be smarter to somehow attach the mapping logic to the UaVariableNode itself but couldn't find a method that is designed to supply runtime data. Is there a best-practise/example for this?

**Comment by kevinherron** [comment]
Not sure if this is exactly what you’re after, but the `AttributeFilter` mechanism allows you to “hook in” and provide dynamic attributes for Nodes:

https://github.com/eclipse/milo/blob/0c67133ef5a597df1bc0ed49d153a702c846820e/milo-examples/server-examples/src/main/java/org/eclipse/milo/examples/server/ExampleNamespace.java#L492-L498

Building a “proper” aggregating server is a pretty difficult undertaking so there is no example. If I were attempting this I would probably not represent each Node in the underlying server with my own Node in memory, but rather, try to pass through each of the service calls on the AddressSpace to the underlying client/server pair. Or at least that’s the first thing I’d investigate as a possibility.

**Reply by arnoweiss** [reply]
Thanks, I'll proceed down the `AttributeFilter `route

**Comment by rafaone** [comment]
> Not sure if this is exactly what you’re after, but the `AttributeFilter` mechanism allows you to “hook in” and provide dynamic attributes for Nodes:
> 
> https://github.com/eclipse/milo/blob/0c67133ef5a597df1bc0ed49d153a702c846820e/milo-examples/server-examples/src/main/java/org/eclipse/milo/examples/server/ExampleNamespace.java#L492-L498
> 
> Building a “proper” aggregating server is a pretty difficult undertaking so there is no example. If I were attempting this I would probably not represent each Node in the underlying server with my own Node in memory, but rather, try to pass through each of the service calls on the AddressSpace to the underlying client/server pair. Or at least that’s the first thing I’d investigate as a possibility.

I'm trying this but in variant value i have a  value from DAO object (Main.thMain.wrl0.lastHistory.getTemp),  and when i update this object the value not refresh. what is the issue?       


```
UaVariableNode node = new UaVariableNode.UaVariableNodeBuilder(getNodeContext())
	                .setNodeId(newNodeId( pNode.getNodeId().getIdentifier() + "." + name))
	                .setAccessLevel(AccessLevel.READ_ONLY)
	                .setBrowseName(newQualifiedName(name))
	                .setDisplayName(LocalizedText.english(name))
	                .setDataType(Identifiers.Integer)
	                .setTypeDefinition(Identifiers.BaseDataVariableType)
	                .setMinimumSamplingInterval(MIN_SAMPLING)
	                .build();
	            node.getFilterChain().addLast(
	                new AttributeLoggingFilter(),
	                AttributeFilters.getValue(
	                    ctx ->
	                        new DataValue(new Variant(Main.thMain.wrl0.lastHistory.getTemp()))
	                )
	            );
	            if(unit!=null){
	            	node.setEngineeringUnits(unit);
	            }
	            getNodeManager().addNode(node);
	            pNode.addComponent(node);
```
Whey a set new value for this object Main.thMain.wrl0.lastHistory.setTemp(1000); its not update in OPCUA explorer its read the initial value when its created.

**Reply by kevinherron** [reply]
Run it in a debugger, place a breakpoint, and see if you're hitting that line each time the Node is read.

If you can't figure out what's happening then start a new discussion. You'll probably need to provide more code and context to troubleshoot further.
